subiquity/ui/views/identity.py

```diff
@@ -95,6 +99,23 @@ def __init__(self, reserved_usernames, initial):
     password = PasswordField(_("Choose a password:"))
     confirm_password = PasswordField(_("Confirm your password:"))
 
+    def on_username_change(self, _, value):
+        if len(value) < 2:
+            return
+
+        if self.validation_task is not None:
+            self.validation_task.cancel()
+
+        self.validation_task = \
+            schedule_task(self.controller.validate_username(value))
+
+    def username_validation_state(self):
+        task = self.validation_task
+        if task is None or not task.done():
+            return UsernameValidation.OK
```

Comment:
Oh boy, asynchronous validation! I certainly wasn't thinking about that when I wrote the forms code.

While this is probably fine in practice, it is theoretically racy to assume that if validation hasn't completed it's OK because you can enter invalid data, leave the field before validation completes and then hit done, which will crash because the validation in POST fails.

Fixing this properly probably involves some delicate hacking of the generic form code -- if the user moves out of a form with validation incomplete, we could do something like freeze the ui for 0.1s and if the validation still has not completed show a spinner and update the validation status when the result comes in. Sounds fiddly though and definitely not something for Friday afternoon!

In this case I think it's probably OK though because the server side of the validation method is not actually asynchronous and won't force you to do this "properly" :)

Reply:
I know you're not a fan of using multiple HTTP requests but I think we can do something like:

when the OK/Done button is pressed, don't submit the form just yet.
* send an asynchronous validation request (through GET).
* display a spinner if needed
* on success, submit the token that was validated (through POST).

To reduce the number of requests, we can also make the POST call return a boolean and make sure we check it on the client-side.

IMO, I don't think we should rely too much on the client-side validation to decide if the application crashes or not :)

Reply:
Well, we should send the validation request when the focus moves out of the field (or when a key is pressed, maybe) rather than when OK/Done is pressed but otherwise that makes sense as a plan I think.

Reply:
It might be worth spending a few minutes thinking about how to handle validation being asynchronous in the generic code. I guess an "if iscoroutine()" somewhere around https://github.com/canonical/subiquity/blob/main/subiquitycore/ui/form.py#L241 would be the start... but I expect that might be a bit painful.

Reply:
> Well, we should send the validation request when the focus moves out of the field (or when a key is pressed, maybe) rather than when OK/Done is pressed but otherwise that makes sense as a plan I think.

That's exactly what I'm doing on the Flutter client. I could not find a suitable signal to connect in the TUI form editor, though.

Reply:
Ah there's no signal but you can define a lost_focus method on a widget and it will get called when the widget loses focus. But the validate_foo method is called by such a hook (https://github.com/canonical/subiquity/blob/main/subiquitycore/ui/form.py#L97) -- so you can start the request there?

Reply:
Ok. I built on top of your suggestion of moving the async calls for when we leave focus. I did https://github.com/canonical/subiquity/pull/1296/commits/df8b157ae891fc441b33ed96a9d0a305eed6da52 as the good-enough approach, but decided to take out the async related logic into a specific mixin class, that can be combined with other editors in https://github.com/canonical/subiquity/pull/1296/commits/e6a6f7a92b2c02eae037e8e580a384940160ec68. Shall we decide to validate hostname in future, there it will be a building block. 

Either of the commits should be fine. I'll leave both for your appreciation. If you decide that the latest is overkill, I'll drop it. Otherwise, I'll be happy to hear from you on the current design.

Reply:
I think the current approach is OK, and can be made more generic as and when we need it. Some test changes required though!